[dtensor] handle negative dim and fix TP regression

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111750

TP style still have some regression due to negative dim specifications,
fix it by allow DTensor API to handle negative dims and normalize them.

i.e. TP uses `Shard(-1)`, and then try to redistribute `Shard(1) -> Shard(-1)`, this should ideally be no-op but current it runs a decompose sharding phrase and it would turn this transformation to `Shard(1) -> Replicate -> Shard(-1)`, which is wrong and triggers unnecessary allgathers

wanchaol committed:
[dtensor] handle negative dim and fix TP regression

TP style still have some regression due to negative dim specifications,
fix it by allow DTensor API to handle negative dims and normalize them

[ghstack-poisoned]

rohan-varma commented on torch/distributed/_tensor/_utils.py:
```diff
@@ -143,7 +143,11 @@ def compute_global_tensor_info(
+            if shard_placement.dim < 0:
+                shard_placement.dim += len(tensor_shape)
+            shard_dim = shard_placement.dim
+            # normalize shard dim to be positive
```
nit: should the comment be a few lines above?

wanchaol replied:
good catch, done

wanchaol committed:
Update on "[dtensor] handle negative dim and fix TP regression"


TP style still have some regression due to negative dim specifications,
fix it by allow DTensor API to handle negative dims and normalize them.

i.e. TP uses `Shard(-1)`, and then try to redistribute `Shard(1) -> Shard(-1)`, this should ideally be no-op but current it runs a decompose sharding phrase and it would turn this transformation to `Shard(1) -> Replicate -> Shard(-1)`, which is wrong and triggers unnecessary allgathers

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wanchaol committed:
Update on "[dtensor] handle negative dim and fix TP regression"


TP style still have some regression due to negative dim specifications,
fix it by allow DTensor API to handle negative dims and normalize them.

i.e. TP uses `Shard(-1)`, and then try to redistribute `Shard(1) -> Shard(-1)`, this should ideally be no-op but current it runs a decompose sharding phrase and it would turn this transformation to `Shard(1) -> Replicate -> Shard(-1)`, which is wrong and triggers unnecessary allgathers

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
